In [1]:
import pandas as pd
import numpy as np

In [2]:
import seaborn as sns
sns.get_dataset_names()

['anagrams',
 'anscombe',
 'attention',
 'brain_networks',
 'car_crashes',
 'diamonds',
 'dots',
 'dowjones',
 'exercise',
 'flights',
 'fmri',
 'geyser',
 'glue',
 'healthexp',
 'iris',
 'mpg',
 'penguins',
 'planets',
 'seaice',
 'taxis',
 'tips',
 'titanic']

In [3]:
len(sns.get_dataset_names())

22

In [6]:
## Load the iris dataset

sns.load_dataset("taxis")

pickup             dropoff  passengers  distance  fare  \
0    2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60   7.0   
1    2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79   5.0   
2    2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37   7.5   
3    2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0   
4    2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16   9.0   
...                  ...                 ...         ...       ...   ...   
6428 2019-03-31 09:51:53 2019-03-31 09:55:27           1      0.75   4.5   
6429 2019-03-31 17:38:00 2019-03-31 18:34:23           1     18.74  58.0   
6430 2019-03-23 22:55:18 2019-03-23 23:14:25           1      4.14  16.0   
6431 2019-03-04 10:09:25 2019-03-04 10:14:29           1      1.12   6.0   
6432 2019-03-13 19:31:22 2019-03-13 19:48:02           1      3.85  15.0   

       tip  tolls  total   color      payment            pickup_zone  \
0     2.15    0.0  12.95  yellow  credit card        Lenox Hill West   
1     0.00    0.0   9.30  yellow         cash  Upper West Side South   
2     2.36    0.0  14.16  yellow  credit card          Alphabet City   
3     6.15    0.0  36.95  yellow  credit card              Hudson Sq   
4     1.10    0.0  13.40  yellow  credit card           Midtown East   
...    ...    ...    ...     ...          ...                    ...   
6428  1.06    0.0   6.36   green  credit card      East Harlem North   
6429  0.00    0.0  58.80   green  credit card                Jamaica   
6430  0.00    0.0  17.30   green         cash    Crown Heights North   
6431  0.00    0.0   6.80   green  credit card          East New York   
6432  3.36    0.0  20.16   green  credit card            Boerum Hill   

                          dropoff_zone pickup_borough dropoff_borough  
0                  UN/Turtle Bay South      Manhattan       Manhattan  
1                Upper West Side South      Manhattan       Manhattan  
2                         West Village      Manhattan       Manhattan  
3                       Yorkville West      Manhattan       Manhattan  
4                       Yorkville West      Manhattan       Manhattan  
...                                ...            ...             ...  
6428              Central Harlem North      Manhattan       Manhattan  
6429  East Concourse/Concourse Village         Queens           Bronx  
6430                    Bushwick North       Brooklyn        Brooklyn  
6431      East Flatbush/Remsen Village       Brooklyn        Brooklyn  
6432                   Windsor Terrace       Brooklyn        Brooklyn  

[6433 rows x 14 columns]

## Load the mpg dataset from seaborn library

In [7]:
df = sns.load_dataset("mpg")
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                       name  
0            70     usa  chevrolet chevelle malibu  
1            70     usa          buick skylark 320  
2            70     usa         plymouth satellite  
3            70     usa              amc rebel sst  
4            70     usa                ford torino  
..          ...     ...                        ...  
393          82     usa            ford mustang gl  
394          82  europe                  vw pickup  
395          82     usa              dodge rampage  
396          82     usa                ford ranger  
397          82     usa                 chevy s-10  

[398 rows x 9 columns]

In [12]:
## Display the starting five rows of the data

df.head()     ## When nothing is specified inside the head() it will always display the first 5 rows

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [10]:
## Display the starting 3 rows

df.head(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite

In [11]:
## Display the starting 10 rows 
df.head(10)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   
5  15.0          8         429.0       198.0    4341          10.0   
6  14.0          8         454.0       220.0    4354           9.0   
7  14.0          8         440.0       215.0    4312           8.5   
8  14.0          8         455.0       225.0    4425          10.0   
9  15.0          8         390.0       190.0    3850           8.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino  
5          70    usa           ford galaxie 500  
6          70    usa           chevrolet impala  
7          70    usa          plymouth fury iii  
8          70    usa           pontiac catalina  
9          70    usa         amc ambassador dpl

In [16]:
## Display the bottom 5 rows of the data

df.tail()    ## When nothing specified inside the tail() it will always display the bottom 5 rows

mpg  cylinders  displacement  horsepower  weight  acceleration  \
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin             name  
393          82     usa  ford mustang gl  
394          82  europe        vw pickup  
395          82     usa    dodge rampage  
396          82     usa      ford ranger  
397          82     usa       chevy s-10

In [14]:
df.tail(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year origin           name  
395          82    usa  dodge rampage  
396          82    usa    ford ranger  
397          82    usa     chevy s-10

In [15]:
df.tail(10)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
388  26.0          4         156.0        92.0    2585          14.5   
389  22.0          6         232.0       112.0    2835          14.7   
390  32.0          4         144.0        96.0    2665          13.9   
391  36.0          4         135.0        84.0    2370          13.0   
392  27.0          4         151.0        90.0    2950          17.3   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                        name  
388          82     usa  chrysler lebaron medallion  
389          82     usa              ford granada l  
390          82   japan            toyota celica gt  
391          82     usa           dodge charger 2.2  
392          82     usa            chevrolet camaro  
393          82     usa             ford mustang gl  
394          82  europe                   vw pickup  
395          82     usa               dodge rampage  
396          82     usa                 ford ranger  
397          82     usa                  chevy s-10

In [17]:
## Display the total rows and columns from the data

df.shape

(398, 9)

### Observation:

- There are 398 rows and 9 columns in the data

In [18]:
## Display all the column names

df.columns

Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'model_year', 'origin', 'name'],
      dtype='object')

In [19]:
## Display only the name column from the data

df["name"]

0      chevrolet chevelle malibu
1              buick skylark 320
2             plymouth satellite
3                  amc rebel sst
4                    ford torino
                 ...            
393              ford mustang gl
394                    vw pickup
395                dodge rampage
396                  ford ranger
397                   chevy s-10
Name: name, Length: 398, dtype: object

In [21]:
## Display the cylinders column 

df[["cylinders"]]

cylinders
0            8
1            8
2            8
3            8
4            8
..         ...
393          4
394          4
395          4
396          4
397          4

[398 rows x 1 columns]

In [22]:
## How many cars having4 cylinder engine and 8 cylinder engine

df["cylinders"].value_counts()

4    204
8    103
6     84
3      4
5      3
Name: cylinders, dtype: int64

### observation:

- Out of 398 cars,
204(51.25%) cars have 4 cylinder engines, 103(25.87%) have 8 cylinder engines, 84(21%) have 6 cylinder engines, 4(1%) have 3 cylinder engines, 3(0%) have 5 cylinder engines.

- Most of the cars in my data are 4 cylinder engine cars.

In [23]:
df["cylinders"].value_counts(normalize = True) * 100

4    51.256281
8    25.879397
6    21.105528
3     1.005025
5     0.753769
Name: cylinders, dtype: float64

In [28]:
## What is the range of model-year of the cars

df["model_year"].unique()

array([70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82], dtype=int64)

In [29]:
## Check the horse power of the cars

df["horsepower"].unique()

array([130., 165., 150., 140., 198., 220., 215., 225., 190., 170., 160.,
        95.,  97.,  85.,  88.,  46.,  87.,  90., 113., 200., 210., 193.,
        nan, 100., 105., 175., 153., 180., 110.,  72.,  86.,  70.,  76.,
        65.,  69.,  60.,  80.,  54., 208., 155., 112.,  92., 145., 137.,
       158., 167.,  94., 107., 230.,  49.,  75.,  91., 122.,  67.,  83.,
        78.,  52.,  61.,  93., 148., 129.,  96.,  71.,  98., 115.,  53.,
        81.,  79., 120., 152., 102., 108.,  68.,  58., 149.,  89.,  63.,
        48.,  66., 139., 103., 125., 133., 138., 135., 142.,  77.,  62.,
       132.,  84.,  64.,  74., 116.,  82.])

In [30]:
## Display the summary report the data

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


In [31]:
## Check the summary statistics of all the object columns

df.describe(include = "O")

origin        name
count     398         398
unique      3         305
top       usa  ford pinto
freq      249           6

In [32]:
df['origin'].unique()

array(['usa', 'japan', 'europe'], dtype=object)

In [33]:
## Get the summary statistics of all the numerical cols

df.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

In [35]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [36]:
## Display name, origin and the model_year

df[["name", "origin", "model_year"]]

name  origin  model_year
0    chevrolet chevelle malibu     usa          70
1            buick skylark 320     usa          70
2           plymouth satellite     usa          70
3                amc rebel sst     usa          70
4                  ford torino     usa          70
..                         ...     ...         ...
393            ford mustang gl     usa          82
394                  vw pickup  europe          82
395              dodge rampage     usa          82
396                ford ranger     usa          82
397                 chevy s-10     usa          82

[398 rows x 3 columns]

## Indexing

In [38]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [44]:
## Display 3rd row(index 2)

df.iloc[2]    ## iloc stands for index location (i = index, loc = location)

mpg                           18.0
cylinders                        8
displacement                 318.0
horsepower                   150.0
weight                        3436
acceleration                  11.0
model_year                      70
origin                         usa
name            plymouth satellite
Name: 2, dtype: object

In [45]:
## Display 3rd row(index 2) and 5th row (index 4)

df.iloc[[2, 4]]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
2  18.0          8         318.0       150.0    3436          11.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                name  
2          70    usa  plymouth satellite  
4          70    usa         ford torino

In [46]:
## Display 100th row(index 99) and 125th row(index 124) and 150th row(index 149)

df.iloc[[99,124,149]]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
99   18.0          6         232.0       100.0    2945          16.0   
124  11.0          8         350.0       180.0    3664          11.0   
149  24.0          4         120.0        97.0    2489          15.0   

     model_year origin              name  
99           73    usa        amc hornet  
124          73    usa  oldsmobile omega  
149          74  japan       honda civic

In [51]:
## Display the weight of 5th row(index 4)

df.iloc[[4], [4]]

weight
4    3449

In [53]:
df.iloc[4, 4]

3449

In [54]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [55]:
## Display weight and name of 50th row(index 49) and 75th row(index 74)

df.iloc[[49, 74], [4, 8]]

weight                   name
49    2220     mercury capri 2000
74    4294  ford gran torino (sw)

In [56]:
## Display horsepower, acceleration and name of 3rd row(index 2) and 8th row(index 7)

df.iloc[[2, 7], [3, 5, 8]]

horsepower  acceleration                name
2       150.0          11.0  plymouth satellite
7       215.0           8.5   plymouth fury iii

In [57]:
df.iloc[[2,7], ['horsepower', 'acceleration', 'name']]

IndexError: .iloc requires numeric indexers, got ['horsepower' 'acceleration' 'name']

In [58]:
df.loc[[2,7], ['horsepower', 'acceleration', 'name']]

horsepower  acceleration                name
2       150.0          11.0  plymouth satellite
7       215.0           8.5   plymouth fury iii

In [59]:
## Display 300th row, 325th row and 350 row of mpg, cylinders and origin

df.loc[[299,324,349],["mpg","cylinders","origin"]]

mpg  cylinders  origin
299  27.2          4  europe
324  40.8          4   japan
349  34.1          4   japan

In [60]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

## NumPy Revisit

In [61]:
## Write a numpy pgm to find the min and max of an array

arr = np.arange(5, 15)
arr

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [62]:
np.min(arr)

5

In [63]:
np.max(arr)

14

In [73]:
## Get the index location of the min value and the max of the array

np.argmin(arr)

0

In [74]:
np.argmax(arr)

9

In [76]:
arr1 = np.random.randint(50, 150, 25)
arr1

array([104, 112, 148,  61,  79, 123,  70, 124, 123,  50,  61,  80, 112,
        59,  89,  71,  83,  55, 102,  65, 118,  75, 100, 134,  97])

In [77]:
np.argmax(arr1)    ## index 2 = 148 i.e. the max value

2

In [79]:
np.argmin(arr1)   ## index 9 = 50 i.e the min value of this array

9

In [80]:
arr1

array([104, 112, 148,  61,  79, 123,  70, 124, 123,  50,  61,  80, 112,
        59,  89,  71,  83,  55, 102,  65, 118,  75, 100, 134,  97])

In [82]:
## Get all the numbers greater than 100

arr1[arr1 > 100]    ## Condition indexing

array([104, 112, 148, 123, 124, 123, 112, 102, 118, 134])

In [84]:
## Get all the numbers greater than 100 and divide those nos by 10

# np.where(condition, name of the array, arithematic operation)

np.where(arr1 > 100, arr1, arr1 / 100)

In [87]:
np.where(arr1 % 5 == 0, arr1, arr1 * 5)    ## Check this

array([520, 560, 740, 305, 395, 615,  70, 620, 615,  50, 305,  80, 560,
       295, 445, 355, 415,  55, 510,  65, 590,  75, 100, 670, 485])

In [86]:
arr1

array([104, 112, 148,  61,  79, 123,  70, 124, 123,  50,  61,  80, 112,
        59,  89,  71,  83,  55, 102,  65, 118,  75, 100, 134,  97])